[INF-0004] Flesh out debug layer bytecode validation behavior. (#290)
microsoft/hlsl-specs: proposals/infra/INF-0004-validator-hashing.md

internal validator.

For pre-release shader models DXC should run the validator, but it should not
apply the hash. Instead it should apply the `PREVIEW_BYPASS` sentinel value 
from the table above. This allows us to differentiate between shaders validated 
by a "final" validator.

### IDxcValidator API Update

A new `IDxcValidator` API flag is needed to avoid container hashing and
unnecessary overhead when used by runtime validation.

Proposed change to flags in `dxcapi.h`:

> ```cpp

Reviewer: If you specify this as `diff` instead of `cpp` the rendering will actually render as an inline unified diff.

Reviewer: Or just strip it down as @tex3d suggested below (I guess I should read the other comments before commenting).

runtime.

> Note: an important behavior change here is that shaders with zero'd hash data
> will always be treated as invalid by the runtime and rejected.
> 
> An exception is for a period of time already existing preview shaders that used 
> zero'd hash data will be treated like having the `PREVIEW_BYPASS` hash in that
> preview runtimes will allow them to be used, at least until those preview shaders
> and runtimes are no longer supported.  In contrast to `PREVIEW_BYPASS` shaders,
> these legacy zero'd has shaders will not run have the validator invoked on them
> by the runtime/debug layer since that wasn't expected of them.

#### D3D Runtime Validation Control

Apps can configure how the runtime validates bytecode passed to it. 

The validation referred to here is implementation exposed by `dxil.dll` 
or `dxcompiler.dll`, and also embedded in `d3d12sdklayers.dll`.  
This is the same validation used when the compiler endorses bytecode at 
compile time by applying a hash.  So it would typically be 
redundant to validate again, but it can be useful to validate shaders that have 
the `BYPASS` or `PREVIEW_BYPASS` hash, or have the option to force validation. 

> If there is bytecode to validate, the runtime and/or debug layer attempt to 

Reviewer 2: > the runtime and/or debug layer

Just to clarify here (because I was under the impression that only the debug layer was ever going to do validation).  Is the proposal here that somethings the runtime will do validation (by calling into some external DLL) but at other times the debug layer will do validation (using a version statically linked with it)?

Or something else?

Author: The flags let the user decide what they want.   All the details are there, but brief summary:
- by default only debug layer validates only BYPASS/BYPASS_PREVIEW shaders, printing messages and not failing. The debug layer doesn't alter pass/fail behavior by design (to the extent possible)
- the app can choose to have the runtime validate as well, with or without debug layer.  When the runtime is validating, it will fail (e.g. state object creation fail) when it sees an error.  So an app could choose this mode if it wants just an extra layer of checking on any shader it gets in the field in retail (e.g. without debug layer)
- the app can choose for the validation to apply to all shaders, even hashed ones, if it wants.   Maybe they want to try a new validator.
- The app can choose where the validator implementation comes from, incl from debug layer (when enabled) or dxil.dll/dxcompiler.dll whichever is present.  The search order can be tweaked.
- The app can choose to skip validation if an implementation isn't available, or have validation fail if an implementation isn't available
- If the debug layer and runtime are both doing validation (runtime for pass/fail, debug layer for printing messages), they will always select the same validator implementation so they are consistent.   This pattern is similar to how a huge amount of existing runtime+debug layer validation works - they are guaranteed to be consistent with each other.

Reviewer 2: I can see that the proposal adds a huge amount of flexibility.  I'm unsure, though, why we're planning on doing any of that and why it is necessary?

Author: This is defining what was originally proposed, which is that there would be controls on when to validate.

If we're sanctioning the workflow of bypassing the validator during author time, the least we should do is let apps validate at runtime.  There isn't any extra burden on HLSL team here since this is validation code the app can also run themselves manually.  But I think it is of great utility to make it straightforward to perform the validation in all the paths and engine ends up feeding shaders to D3D by simply having the runtime do it when it sees a shader (vs the app having to locate all the paths where they might be sourcing shaders from.

Author: Per offline discussion, removed runtime validation support.

For shaders with the `BYPASS` hash, shaders can run without validating the hash 
regardless of whether the machine is in developer mode as long as the shader 
targets a supported non-experimental shader model.

All other values of the hash are assumed to be real hash values.

For shaders that contain a real hash value, the behavior is unchanged. The hash
is verified and the shader executes as it does with existing versions of the 
runtime.

> Note: an important behavior change here is that shaders with zero'd hash data
> will always be treated as invalid by the runtime and rejected.
> 
> An exception is for a period of time already existing preview shaders that used 
> zero'd hash data will be treated as equivalent to having the `PREVIEW` hash 
> until those preview shaders are no longer supported in new preview runtimes.

#### D3D Runtime Validation Control

Apps can configure how the runtime validates bytecode passed to it. 

The validation referred to here is implementation exposed by `dxcompiler.dll` 
or `dxil.dll`.  This is the same validation used when the compiler
endorses bytecode at compile time by applying a hash.  So it would typically be 

> redundant to validate again  But it can be useful to validate shaders that have 

Reviewer: ```suggestion
redundant to validate again, but it can be useful to validate shaders that have 
```

> the `BYPASS` hash, or have the option to force validation for whatever reason. 

Reviewer: nit: unnecessarily wordy
```suggestion
the `BYPASS` hash, or have the option to force validation. 
```

`D3D12_BYTECODE_FLAG_SKIP_VALIDATION_IF_VALIDATOR_NOT_AVAILABLE` | Regardless of how the other flags request validation, if the runtime or debug layer can't load `dxcompiler.dll` or `dxil.dll` and find a validation implementation they skip validation without failing.  If this flag is not set, the runtime will fail if it needs to validate bytecode and can't find a validator implementation.
`D3D12_BYTECODE_FLAGS_DEFAULT` | By default validation is only done on bytecode with the `BYPASS` hash, only from the debug layer, and only if a validator implementation can be found.  Equivalent to the three flags: `D3D12_BYTECODE_FLAG_VALIDATE_BYTECODE_WITH_BYPASS_HASH`, `D3D12_BYTECODE_FLAG_VALIDATE_ONLY_IF_DEBUG_LAYER_ENABLED_NO_FAILING` and `D3D12_BYTECODE_FLAG_SKIP_VALIDATION_IF_VALIDATOR_NOT_AVAILABLE` |

The first three flags above are mutually exclusive and one must be chosen.

Flags are set via `ID3D12BytecodeOptions`:

```C++
[uuid(9291ad45-9a10-43a5-9c06-f5aa558e89a7), object, local, pointer_default(unique)]
interface ID3D12BytecodeOptions
    : IUnknown
{
    HRESULT SetBytecodeFlags(
        [annotation("_In_")] D3D12_BYTECODE_FLAGS flags

>     );

Author: Should there be any other valdiation controls?  Like allowing behavior tweaks on the validator in some way? (e.g validator version setting?)

Author: Ok, so we need HLSL work item to add this no-hash option.

Reviewer: I'm not sure I follow why we would add a no-hash option. Shouldn't we either run the validator and hash or not run the validator and thus not hash?

Author: This is for when the runtime calls the validator.  It's just interested in whether the bytecode is valid or not, no need to hash.

Reviewer: I still question the need for this. Don't we have a flag to write in place? That should avoid increased memory impact, and if we're not writing the output we just wrote some memory that gets thrown away.

I'm unsure that the cost of calculating the hash for something that happens only in the debug layer is actually worth the extra support complexity of a new user exposed interface option.

Author: There is an  DxcValidatorFlags_InPlaceEdit flag.  The runtime could try to make sure it's only ever calling validate with it's own copy of bytecode, never passing in the APIs caller's memory.  But even if it did that, the runtime would lose track of what BYPASS the shader had originally since it would now be stomped by an unneeded hash.  Runtime could cache the bypass state before calling validator.  But it would seem cleanest overall if it could just ask for validation only without side effects.

Reviewer: Caching the bypass state before calling the validator is less clean than calling the validator handling a potential error flag then needing to cache the bypass state anyways and call the validator without the flag?

It seems like whatever you need to do in the runtime to make this work with existing validators needs to be done regardless of whether or not we add a flag to the validator.

Author: For existing validators we'd just use the default behavior which is the validator gives us a copy of the shader that's hashed.

Much cleaner if we can first just try to validate with the new flag, which will avoid making a copy of the shader.  If the call fails E_INVALIDARG, since the new flag isn't recognized, we just call again without the flag and live with the fact that the validator made an unneeded copy of the shader.

Author: My earlier reply was exploring what we'd have to do if we used the in-place flag that you pointed out.  What we'd really do is above.  Easy way to avoid wasted shader copy if possible.  If the request isn't reasonable, then what I list as the fallback just has to be the only path.

Reviewer: I think that if the fallback path needs to be implemented anyways, we should start there and only add the extra flag once we can measure the performance.

That allows us to understand the impact and make an informed decision about prioritization. Every hour we spend working on DXC is an hour we don't spend on Clang, and every new feature we add to DXC is a new feature we have to figure out how to support and/or deprecate as we move forward.

Reviewer 3: > I'm unsure that the cost of calculating the hash for something that happens only in the debug layer is actually worth the extra support complexity of a new user exposed interface option.

Note: this will be called from normal runtime, not just from debug layer, when using `D3D12_BYTECODE_FLAG_VALIDATE_WHEN_HASH_BYPASSED` or `D3D12_BYTECODE_FLAG_VALIDATE_ALL_BYTECODE` flags.

Author: I can cut this flag from consideration - would you prefer that I just delete the section from the spec, or leave it in but labelled with "This section is cut until D3D team can prove that avoiding an unnecessary shader copy really matters."

> Note: an important behavior change here is that shaders with zero'd hash data
> will always be treated as invalid by the runtime and rejected.
> 
> An exception is for a period of time already existing preview shaders that used 
> zero'd hash data will be treated as equivalent to having the `PREVIEW` hash 
> until those preview shaders are no longer supported in new preview runtimes.

#### D3D Runtime Validation Control

Apps can configure how the runtime validates bytecode passed to it. 

The validation referred to here is implementation exposed by `dxcompiler.dll` 
or `dxil.dll`.  This is the same validation used when the compiler
endorses bytecode at compile time by applying a hash.  So it would typically be 
redundant to validate again, but it can be useful to validate shaders that have 
the `BYPASS` hash, or have the option to force validation. 

If there is bytecode to validate, the runtime and/or debug layer attempt to 
load one of these dlls, `dxcompiler.dll` first. If either is available, 
bytecode validation is possible. 

The following flags in `d3d12.h` show the validation control options.  
`ID3D12Device` exposes `ID3D12BytecodeOptions` to set the flags, shown 

> further below.

Reviewer 3: I feel like the API details should be placed under a new `## Detailed Design` section. immediately preceding the `## Appendix 1: DXIL Hashing Algorithm` section.  In this section, the need for, and expected functionality, of the proposed interface could be discussed.

table above. This allows us to differentiate between shaders validated by a
"final" validator.

### IDxcValidator API Update

A new `IDxcValidator` API flag is needed to avoid container hashing and
unnecessary overhead when used by runtime validation.

Proposed change to flags in `dxcapi.h`:

```cpp
 static const UINT32 DxcValidatorFlags_Default = 0;
 static const UINT32 DxcValidatorFlags_InPlaceEdit =
    1; // Validator is allowed to update shader blob in-place.
 static const UINT32 DxcValidatorFlags_RootSignatureOnly = 2;
 static const UINT32 DxcValidatorFlags_ModuleOnly = 4;
+static const UINT32 DxcValidatorFlags_SkipHash =
+    8; // Skip hash update; no modification or copy of input blob.
 static const UINT32 DxcValidatorFlags_ValidMask = 0xF;

> ```

Reviewer 3: Since this was confusing as a diff (which wasn't even entirely correct for the mask change), maybe we should only list the new flag.

Reviewer 3: ```suggestion
Propose adding `DxcValidatorFlags_SkipHash = 8` flag to `dxcapi.h`.  This will indicate to the validator that container hashing is not requested, avoiding an extra copy of the shader and the hashing operation, which is not needed for runtime or sdklayers use cases.
```

> The runtime attempts to use the `DxcValidatorFlags_ValidMask` flag,

Reviewer 3: This should have been the `DxcValidatorFlags_SkipHash` flag, which is the newly added one.

```suggestion
The runtime attempts to use the `DxcValidatorFlags_SkipHash` flag,
```

title. The high rate of bugs in preview compilers contributes to making this
even less likely.

## Detailed Design

### D3D Runtime Behavior

Starting with AgilitySDK version TBD, the runtime adopts new behavior for how 
and when the shader hash is validated, including support for bytecode validation 
in the debug layer.

For shaders with the `PREVIEW_BYPASS` hash, the shader is allowed to execute 
only if developer mode and the experimental feature 
`D3D12ExperimentalShaderModels` is enabled, otherwise the runtime will produce
an error.

For shaders with the `BYPASS` hash, shaders can run without validating the hash 
regardless of whether the machine is in developer mode as long as the shader 
targets a supported non-experimental shader model.

All other values of the hash are assumed to be real hash values.

For shaders that contain a real hash value, the behavior is unchanged. The hash
is verified and the shader executes as it does with existing versions of the 
runtime.

> Note: an important behavior change here is that shaders with zero'd hash data
> will always be treated as invalid by the runtime and rejected.
> 

> > An exception is for a period of time already existing preview shaders that used 
> > zero'd hash data will be treated like having the `PREVIEW_BYPASS` hash in that
> > preview runtimes will allow them to be used, at least until those preview shaders
> > and runtimes are no longer supported.  In contrast to `PREVIEW_BYPASS` shaders,
> > these legacy zero'd has shaders will not run have the validator invoked on them
> > by the runtime/debug layer since that wasn't expected of them.

Reviewer 2: I think this section might have been overly edited, I'm not exactly sure what some of it is trying to say.  An attempt at clarifying:

```suggestion
> We have an exception to allow for legacy zero'd hash shaders. These were built
> using earlier previews and so they're treated as if they had the `PREVIEW_BYPASS` hash set,
> except that the validator will not be invoked on them since this wasn't expected at the time
> they were generated. This exception will be removed once runtimes that support these preview 
> shaders are no longer supported.
```

#### D3D Debug Layer Validation Control

Apps can configure how the debug layer runtime validates bytecode passed to it. 

The validation is the same code the compiler uses when it endorses bytecode at 
compile time by applying a hash.  A copy of this validation code is in 
the debug layer for it to use.  So it would typically be 
redundant to validate again, but it can be useful to validate shaders that have 
the `BYPASS` or `PREVIEW_BYPASS` hash, or have the option to force validation. 

If the debug layer finds errors when running the bytecode validator it only 

> reports messages, but doesn't fail shader creation.

Reviewer 2: Checking here that we don't want to enable a mode where shader creation fails if it fails validation?

Author: This is what the runtime validation option I had was for. The debug layer / runtime interaction is typically it is the runtime that does the failing, and the debug layer's role is just an observer giving commentary.

Gets messy to have to make sure all codepaths work cleanly if the debug layer suddenly starts failing on behalf of the runtime.

I'd be fine bringing back the option for the runtime to fail, as long as debug layer is on.

If we bring this back, I'd switch the enum back to a flags field, so the fail on error can be layered on.

Jesse's argument is an app can write 20 lines of code if they want, to intercept the debug errors and fail if they want.  Certainly an unusual burden to put on them.

Reviewer 2: Or I guess they could just configure the debug layer to break on that error message as well.

I thought I'd run into cases where enabling the debug layer would cause some API calls to fail that didn't previously.

Author: Yes, we try to avoid situations where having the debug layer on changes behavior.  So I started here by proposing that you could ask for the runtime to fail independent of whether the debug layer is on or not.    But based on everyone's hesitation against having the runtime to validation backed off.

Yes, folks can configure the debug layer to break on the error.  Not quite as easy as it seems because there are tons of different codepaths where validation will be invoked from, and we'd have to try to guarantee that they always use the same error ID so folks can break on all possible instances of the failure.  There are some code patterns in the debug layer where there are error ID name specific to a shader stage, for instance, and those code paths would have to make an exception for this particular error.

Author: I will try to do this when I implement.  If when the dust settles I'm confident it's just one ID to break on, I can update this spec to list that ID.

The validation mode can be configured via ID3D12DebugDevice1::SetDebugParameter()`:

```C++
typedef enum D3D12_DEBUG_DEVICE_PARAMETER_TYPE
{
    ...
    D3D12_DEBUG_DEVICE_PARAMETER_BYTECODE_VALIDATION_MODE   
} D3D12_DEBUG_DEVICE_PARAMETER_TYPE;

> typedef enum D3D12_BYTECODE_VALIDATE_MODE

Reviewer 2: For other values:

* `D3D12_DEBUG_DEVICE_PARAMETER_FEATURE_FLAGS` --> `D3D12_DEBUG_FEATURE`
* `D3D12_DEBUG_DEVICE_PARAMETER_GPU_BASED_VALIDATION_SETTINGS` --> `D3D12_DEBUG_DEVICE_GPU_BASED_VALIDATION_SETTINGS`
* `D3D12_DEBUG_DEVICE_PARAMETER_GPU_SLOWDOWN_PERFORMANCE_FACTOR` --> `D3D12_DEBUG_DEVICE_GPU_SLOWDOWN_PERFORMANCE_FACTOR`

Following this pattern, I think the enum should be called:

`D3D12_DEBUG_DEVICE_BYTECODE_VALIDATE_MODE` or `D3D12_DEBUG_BYTECODE_VALIDATE_MODE`

Author: Fixed

> these legacy zero'd has shaders will not run have the validator invoked on them
> by the runtime/debug layer since that wasn't expected of them.

#### D3D Debug Layer Validation Control

Apps can configure how the debug layer runtime validates bytecode passed to it. 

The validation referred to here is implementation exposed by `dxil.dll` 
or `dxcompiler.dll`, and also embedded in the debug layer (`d3d12sdklayers.dll`).
The implementation in the debug layer is used by default, but the external ones 
can be requested.

This is the same validation used when the compiler endorses bytecode at 
compile time by applying a hash.  So it would typically be 
redundant to validate again, but it can be useful to validate shaders that have 
the `BYPASS` or `PREVIEW_BYPASS` hash, or have the option to force validation. 

If the debug layer finds errors when running the bytecode validator it only 
reports messages, but doesn't fail shader creation.

The following flags in `d3d12.h` show the validation control options.  

> `ID3D12Device` exposes `ID3D12BytecodeOptions` to set the flags, shown 

Reviewer 2: Should this configuration be set off the debug interface now, rather than the device?

`D3D12_BYTECODE_FLAG_VALIDATION_DISABLED` | Never invoke bytecode validation.
`D3D12_BYTECODE_FLAG_VALIDATE_WHEN_HASH_BYPASSED` | Only validate bytecode that has the `BYPASS` or `PREVIEW_BYPASS` hash.
`D3D12_BYTECODE_FLAG_VALIDATE_ALL_BYTECODE` | Validate all bytecode regardless of hash.  Forcing validation this way could be useful if there are shaders that are hashed (implying validation ran on them), but there is an updated validator that might catch issues were been missed.  Or the app might want validation on shaders that might have been hashed manually without the compiler's validator.

> `D3D12_BYTECODE_FLAG_PREFER_EXTERNAL_VALIDATOR_IMPLEMENTATION` | With this flag the order the debug layer search for validator implementations is: `dxil.dll`, `dxcompiler.dll`, finally it's own implementation in `d3d12sdklayers.dll`.  In the absence of this flag (default), the debug layer just uses its own implementation.

Reviewer 2: This makes me uneasy.  Anything involving searching for executable code and blindly executing it seems like it is ripe for exploitation.

I guess if it is only enabled for developer machines, or if the debug layer has been loaded, then it's less of an issue.

I don't understand the motivation for this feature at all.
